ext/drpttracer/drpttracer.h

```diff
@@ -0,0 +1,238 @@
+    /** Operation failed: invalid output. */
+    DRPTTRACER_ERROR_INVALID_OUTPUT,
+    /** Operation failed: failed to open perf event. */
+    DRPTTRACER_ERROR_FAILED_TO_OPEN_PERF_EVENT,
+    /** Operation failed: failed to create tracer handle. */
+    DRPTTRACER_ERROR_FAILED_TO_CREATE_PTTRACER_HANDLE,
+    /** Operation failed: failed to start tracing. */
+    DRPTTRACER_ERROR_FAILED_TO_START_TRACING,
+    /** Operation failed: failed to stop tracing. */
+    DRPTTRACER_ERROR_FAILED_TO_STOP_TRACING
+} drpttracer_status_t;
+
+/**
+ * The tracing modes for drpttracer_start_tracing().
+ */
+typedef enum {
+    /** If drpttracer is set to work in this mode, it will trace in userspace. */
+    DRPTTRACER_TRACING_ONLY_USER,
+    /** If drpttracer is set to work in this mode, it will trace in kernel. */
+    DRPTTRACER_TRACING_ONLY_KERNEL,
+    /**
+     * If drpttracer is set to work in this mode, it will trace in both userspace and
+     * kernel.
+     */
+    DRPTTRACER_TRACING_USER_AND_KERNEL
+} drpttracer_tracing_mode_t;
+
+DR_EXPORT
+/**
+ * Starts PT tracing. Must be called after drpttracer_init() and before drpttracer_exit().
+ *
+ * \param[in] drcontext  The context of DynamoRIO.
+ * \param[in] mode  The tracing mode.
+ * \param[in] pt_size_shift  The size shift of PT trace's buffer.
+ * \param[in] sideband_size_shift  The size shift of sideband data's buffer.
+ * \param[out] tracer_handle  The tracer handle.
+ *
+ * \note The size shift is used to control the size of buffers:
+ *       sizeof(PT trace's buffer) = 2 ^ pt_size_shift * PAGE_SIZE.
+ *       sizeof(Sideband data's buffer) = 2 ^ sideband_size_shift * PAGE_SIZE.
+ *
+ * \note The PT trace's buffer and sideband data's buffer are all ring buffers. So Intel
+ * PT will overwrite the oldest data when the buffer is full.
+ *
+ * \note Each tracing corresponds to a trace_handle. When calling
+ * drpttracer_start_tracing(), the caller will get a tracer_handle. And the caller can
+ * stop tracing by passing it to drpttracer_end_tracing().
+ *
+ * \return the status code.
+ */
+drpttracer_status_t
+drpttracer_start_tracing(IN void *drcontext, IN drpttracer_tracing_mode_t mode,
+                         IN uint pt_size_shift, IN uint sideband_size_shift,
+                         OUT void **tracer_handle);
+
+DR_EXPORT
+/**
+ * Stops PT tracing.
+ *
+ * \param[in] drcontext  The context of DynamoRIO.
+ * \param[in] tracer_handle  The tracer handle.
+ * \param[out] output  The output of PT tracing. This function will allocate
+ * and fill this output. It will copy the PT trace to the pt_buff of the output. When the
+ * tracing mode is #DRPTTRACER_TRACING_ONLY_USER or #DRPTTRACER_TRACING_USER_AND_KERNEL,
+ * the PT sideband data will be copy to the sideband_buf of the output. When the tracing
+ * mode is #DRPTTRACER_TRACING_ONLY_KERNEL, it will not copy the sideband data to the
+ * buffer.
+ *
+ * \note The output data can be dumped into different files. After offline tracings are
```

Comment:
The caller can dump the output data to files.

Reply:
Done.